### PR TITLE
sql: retry more distributed errors as local

### LINKED ISSUE #106537: Sql Server Fails to Start with DistSQL Error
**Describe the problem**
SQL servers identify themselves in the system.sql_instances table. The table
contains the sql address and rpc address of the sql server. Here are example 
contents of the sql_instances table:

```
test@fresh-tenant-ver-5005.6h5c.crdb.io:26257/defaultdb> select * from system.sql_instances;
  id |                      addr                      |                session_id                |              locality               |                    sql_addr                    | crdb_region | binary_version
-----+------------------------------------------------+------------------------------------------+-------------------------------------+------------------------------------------------+-------------+-----------------
   1 | 10-0-3-139.us-central1.pod.cluster.local:26257 | \x02c1a251a76346fc91267ee61c139e38       | {"Tiers": "region=gcp-us-central1"} | NULL                                           | \x80        | NULL
   2 | 10-0-5-139.us-central1.pod.cluster.local:26257 | \x010180a704ae5254ce4c048c96b79f2f930dfb | {"Tiers": "region=gcp-us-central1"} | 10-0-5-139.us-central1.pod.cluster.local:26257 | \x80        | 23.1

```

If the sql_instance table contains a stale row with an ipaddress+port that is
reused by something else, then sql servers will fail to start with a distsql
error. It looks like a query run during sql server start up attempts to
schedule a dist sql flow on the stale instance and the distsql query failing
causes the process to crash.

This was discovered while attempting to deflake
https://github.com/cockroachdb/cockroach/issues/105402. The bug could happen in
production if an ip address was reused by another tenant or if a sql server
container crashed in a pod and was restarted.

**To Reproduce**
[PR 106538](https://github.com/cockroachdb/cockroach/pull/106538) contains a regression test. Running the test will fail with the following stack trace:

```
--- FAIL: TestStartTenantWithStaleInstance (9.74s)
    test_server_shim.go:467: migration-job-find-already-completed: failed to connect to n1 at 127.0.0.1:46163: initial connection heartbeat failed: grpc: connection error: desc = "transport: authentication handshake failed: context deadline exceeded" [code 14/Unavailable]
        (1) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).execInternal.func1.1
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:1032
          | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next.func1
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:477
          | github.com/cockroachdb/cockroach/pkg/sql.(*rowsIterator).Next
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:528
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).queryInternalBuffered
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:642
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:694
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:680
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow
          |     github.com/cockroachdb/cockroach/pkg/sql/internal.go:664
          | github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable.CheckIfMigrationCompleted
          |     github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable/migrations_table.go:118
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades.func2
          |     github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:220
          | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
          |     github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades
          |     github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:216
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart
          |     github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1627
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).PreStart
          |     github.com/cockroachdb/cockroach/pkg/server/tenant.go:746
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).Start
          |     github.com/cockroachdb/cockroach/pkg/server/tenant.go:913
          | github.com/cockroachdb/cockroach/pkg/server.(*TestServer).StartTenant
          |     github.com/cockroachdb/cockroach/pkg/server/testserver.go:1188
          | github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartTenant
          |     github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:465
          | github.com/cockroachdb/cockroach/pkg/ccl/serverccl.TestStartTenantWithStaleInstance
          |     github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:463
          | [...repeated from below...]
        Wraps: (2) migration-job-find-already-completed
        Wraps: (3) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
          |     github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:181
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).Dial
          |     github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:103
          | github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run
          |     github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:112
          | github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1
          |     github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:152
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
          |     github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484
          | runtime.goexit
          |     GOROOT/src/runtime/asm_amd64.s:1594
        Wraps: (4) failed to connect to n1 at 127.0.0.1:46163
        Wraps: (5) forced error mark
          | "originated at breaker breaker"
          | github.com/cockroachdb/cockroach/pkg/util/circuit/*circuit.breakerErrorMark::
        Wraps: (6) forced error mark
          | "breaker open"
          | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        Wraps: (7) initial connection heartbeat failed
        Wraps: (8) grpc: connection error: desc = "transport: authentication handshake failed: context deadline exceeded" [code 14/Unavailable]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *markers.withMark (6) *markers.withMark (7) *netutil.InitialHeartbeatFailedError (8) *status.Error
```

Jira issue: CRDB-29607

### LINKED ISSUE #108152: server: TestServerShutdownReleasesSession failed
server.TestServerShutdownReleasesSession [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11185783?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/11185783?buildTab=artifacts#/) on master @ [886515ffc2f1485f47bc3448db93abe3f0217f03](https://github.com/cockroachdb/cockroach/commits/886515ffc2f1485f47bc3448db93abe3f0217f03):

```
=== RUN   TestServerShutdownReleasesSession
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/7f3306bd25c63f593c1cefe649bebfca/logTestServerShutdownReleasesSession3796834340
    test_log_scope.go:81: use -show-logs to present logs inline
    drain_test.go:317: error executing 'SELECT session_id FROM system.sqlliveness WHERE session_id = $1': pq: rpc error: code = Canceled desc = grpc: the client connection is closing
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7f3306bd25c63f593c1cefe649bebfca/logTestServerShutdownReleasesSession3796834340
--- FAIL: TestServerShutdownReleasesSession (3.24s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestServerShutdownReleasesSession.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-30331

### LINKED ISSUE #108271: pkg/ccl/testccl/sqlccl/sqlccl_test: TestTenantTempTableCleanup failed
pkg/ccl/testccl/sqlccl/sqlccl_test.TestTenantTempTableCleanup [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11214351?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11214351?buildTab=artifacts#/) on master @ [f1623e0a2cffc81580ee3c3b71035efd854d89c9](https://github.com/cockroachdb/cockroach/commits/f1623e0a2cffc81580ee3c3b71035efd854d89c9):

```
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowExWithCols
          | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:694
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRowEx
          | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:680
          | github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).QueryRow
          | 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:664
          | github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable.CheckIfMigrationCompleted
          | 	github.com/cockroachdb/cockroach/pkg/upgrade/migrationstable/migrations_table.go:118
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades.func2
          | 	github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:220
          | github.com/cockroachdb/cockroach/pkg/util/startup.RunIdempotentWithRetryEx[...]
          | 	github.com/cockroachdb/cockroach/pkg/util/startup/retry.go:142
          | github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager.(*Manager).RunPermanentUpgrades
          | 	github.com/cockroachdb/cockroach/pkg/upgrade/upgrademanager/manager.go:216
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServer).preStart
          | 	github.com/cockroachdb/cockroach/pkg/server/server_sql.go:1601
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).PreStart
          | 	github.com/cockroachdb/cockroach/pkg/server/tenant.go:754
          | github.com/cockroachdb/cockroach/pkg/server.(*SQLServerWrapper).Start
          | 	github.com/cockroachdb/cockroach/pkg/server/tenant.go:921
          | github.com/cockroachdb/cockroach/pkg/server.(*testServer).StartTenant
          | 	github.com/cockroachdb/cockroach/pkg/server/testserver.go:1507
          | github.com/cockroachdb/cockroach/pkg/testutils/serverutils.StartTenant
          | 	github.com/cockroachdb/cockroach/pkg/testutils/serverutils/test_server_shim.go:277
          | pkg/ccl/testccl/sqlccl/sqlccl_test_test.TestTenantTempTableCleanup
          | 	pkg/ccl/testccl/sqlccl/sqlccl_test_test/pkg/ccl/testccl/sqlccl/temp_table_clean_test.go:163
          | [...repeated from below...]
        Wraps: (2) migration-job-find-already-completed
        Wraps: (3) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).dial
          | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:181
          | github.com/cockroachdb/cockroach/pkg/rpc/nodedialer.(*Dialer).Dial
          | 	github.com/cockroachdb/cockroach/pkg/rpc/nodedialer/nodedialer.go:103
          | github.com/cockroachdb/cockroach/pkg/sql.runnerRequest.run
          | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:112
          | github.com/cockroachdb/cockroach/pkg/sql.(*runnerCoordinator).init.func1
          | 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:152
          | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
          | 	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:484
          | runtime.goexit
          | 	GOROOT/src/runtime/asm_amd64.s:1594
        Wraps: (4) failed to connect to n1 at 127.0.0.1:34173
        Wraps: (5) initial connection heartbeat failed
        Wraps: (6) grpc: client requested node ID 1 doesn't match server node ID 3 [code 2/Unknown]
        Error types: (1) *withstack.withStack (2) *errutil.withPrefix (3) *withstack.withStack (4) *errutil.withPrefix (5) *netutil.InitialHeartbeatFailedError (6) *status.Error
I230807 08:09:21.697240 63957 2@rpc/peer.go:537  [T10,nsql3,rnode=3,raddr=127.0.0.1:34173,class=default,rpc] 2090  connection is now healthy
W230807 08:09:21.723780 63817 sql/temporary_schema.go:466  [T10,nsql3] 2091  error during schema cleanup, retrying: fan out rpc failed with failed to dial into node 1 (NODE_STATUS_LIVE): initial connection heartbeat failed: grpc: client requested node ID 1 doesn't match server node ID 3 [code 2/Unknown] on node 1
E230807 08:09:21.727402 63960 2@rpc/peer.go:611  [T10,nsql3,rnode=1,raddr=127.0.0.1:34173,class=default,rpc] 2092  failed connection attempt (last connected 33ms ago): grpc: client requested node ID 1 doesn't match server node ID 3 [code 2/Unknown]
--- FAIL: TestTenantTempTableCleanup (14.39s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantTempTableCleanup.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-30404

### PR DESCRIPTION
**Pull request by yuzefovich**
This PR contains a couple of commits that increase the allow-list of errors that are retried locally. In particular, it allows us to hide some issues we have around using DistSQL and shutting down SQL pods.

Fixes: #106537.
Fixes: #108152.
Fixes: #108271.

Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/sql/distsql_running.go` line 107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbH6kBo0zrtwF0yaTKa:-NbH6kBo0zrtwF0yaTKb:b-60aso5) ([raw file](https://github.com/cockroachdb/cockroach/blob/8c2c676c74ad2d61b94ace8327fc263bf08b6e6a/pkg/sql/distsql_running.go#L107)):*
> ```Go
> }
> 
> func isDialErr(err error) bool {
> ```

~I briefly looked into using `errors.Is` or `errors.As` here but wasn't sure what's the canonical way that would definitely work. Probably Raphael has a suggestion here.~ Reimplemented with `errors.HasType`.

**Review by yuzefovich — COMMENTED** [review]
I added another commit to include more errors in the allow-list for retrying. Done pushing for now, PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbI7-Z8D_j__vL5L-uJ:b8zh4l0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @mgartner)

**Review by mgartner — APPROVED** [review]
:lgtm: 

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbKFJiRANWT0x1acqZH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson, @knz, and @yuzefovich)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKEthb0p5pVF5sanb7:b2kjo9p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
> ```Go
> 	}()
> 
> 	listener, err := net.Listen("tcp", rpcAddr)
> ```

Why is this listener required?

**Review by jeffswenson — APPROVED** [review]
LGTM

File: pkg/sql/distsql_running.go [hunk 1]
```diff
@@ -100,6 +101,22 @@ type runnerResult struct {
 	err    error
 }
 
+type runnerDialErr struct {
```
Comment (JeffSwenson) [inline pkg/sql/distsql_running.go:104]: I like the approach of tagging errors that occur during the dial flow. This is a robust way to handle issues that occur during start up, which is the most common case we have seen for Serverless.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson and @knz)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKqqpGAx-4DglyikTr:bcoy4qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why is this listener required?
</blockquote></details>

I'm not sure, I just copied this test from Jeff #106538. @JeffSwenson can you share some context about this listener?

**Comment by JeffSwenson** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)_ status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @JeffSwenson and @knz)
> 
> _[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbKqqpGAx-4DglyikTr:bcoy4qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):_
> 
> Previously, mgartner (Marcus Gartner) wrote…
> I'm not sure, I just copied this test from Jeff #106538. @JeffSwenson can you share some context about this listener?

The test is covering the following scenario:
1. A sql server starts up and is assigned port `a`
2. The sql server shuts down and releases port `a`
3. Something else starts up and claims port `a`. In the test that is the listener. This is important because the listener causes connections to `a` to hang instead of responding with a RESET packet.
4. A different server with stale instance information schedules a distsql flow and attempts to dial port `a`.

**Review by yuzefovich — COMMENTED** [review]
Usage of the string matching for errors in `IsDistSQLRetryable` is not great, but it's not a new problem and we do have #82847 as a tracking issue to improve the situation.

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbMLnXiE1W_J2vFKvwb:b-nvw0p8)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbMLhvM0T3pl5mw1aLT:b-hm7rxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

The test is covering the following scenario:
1. A sql server starts up and is assigned port `a`
2. The sql server shuts down and releases port `a`
3. Something else starts up and claims port `a`. In the test that is the listener. This is important because the listener causes connections to `a` to hang instead of responding with a RESET packet.
4. A different server with stale instance information schedules a distsql flow and attempts to dial port `a`.

</blockquote></details>

Thanks for this context. I incorporated it as a comment on the test.

**Comment by yuzefovich** [comment]
Hm, I'm confused
```
[23:12:22][Run unit tests] ERROR: /go/src/github.com/cockroachdb/cockroach/pkg/jobs/BUILD.bazel:88:8: GoCompilePkg pkg/jobs/autoconfig.recompile1376.a failed: (Exit 1): builder failed: error executing command (from target //pkg/jobs:jobs_test) bazel-out/k8-opt-exec-2B5CBBC6/bin/external/go_sdk/builder_reset/builder compilepkg -sdk external/go_sdk -installsuffix linux_amd64 -tags bazel,gss,bazel,gss -src pkg/server/autoconfig/auto_config.go ... (remaining 62 arguments skipped)
[23:12:22][Run unit tests] 
[23:12:22][Run unit tests] Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
[23:12:22][Run unit tests] compilepkg: panic: runtime error: invalid memory address or nil pointer dereference
[23:12:22][Run unit tests] [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x5e89bc]
```
Let's try one more time.

bors r+

**Comment by yuzefovich** [comment]
Known failure #108340.

bors r+

**Comment by yuzefovich** [comment]
The latest failure is at least legitimate:
```
=== RUN   TestStartTenantWithStaleInstance
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
    test_log_scope.go:81: use -show-logs to present logs inline
    server_sql_test.go:450: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
        	Error:      	Received unexpected error:
        	            	listen tcp 127.0.0.1:45845: bind: address already in use
        	Test:       	TestStartTenantWithStaleInstance
```
I'll take a look tomorrow, but I did stress this new test on the gceworker with no failures.

**Comment by mgartner** [comment]
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbQZGDICh6DoJx3ilay:bmgfde0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks for this context. I incorporated it as a comment on the test.
</blockquote></details>

Thanks for the context. The only part I'm missing is how the "different server" is dialing `rpcAddr` - I don't see how that's forced.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbSD8Bq7gOsLlguXVtM:bzh9mgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):*
> ```Go
> 
> 	listener, err := net.Listen("tcp", rpcAddr)
> 	require.NoError(t, err)
> ```

For the failure that I got under stress on CI on this line:
```
=== RUN   TestStartTenantWithStaleInstance
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
    test_log_scope.go:81: use -show-logs to present logs inline
    server_sql_test.go:450: 
        	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
        	Error:      	Received unexpected error:
        	            	listen tcp 127.0.0.1:45845: bind: address already in use
        	Test:       	TestStartTenantWithStaleInstance
```
it seems like the stopped tenant hasn't released the socket, and to me it seems like a benign error with the test setup. I'm inclined to introduce SucceedsSoon until `net.Listen` doesn't return an error. @JeffSwenson WDYT?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 441 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbKEthb0p5pVF5sanb6:-NbSGeRO3hhdz09uXQ05:bc5bdmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbaeb23d3df1a01352057b0df51f74144f7fce33/pkg/ccl/serverccl/server_sql_test.go#L441)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Thanks for the context. The only part I'm missing is how the "different server" is dialing `rpcAddr` - I don't see how that's forced.
</blockquote></details>

I think I see this. In step 4. the "different server" is the new being-started-right-now SQL pod (which itself will listen on a different port), and when starting up, it sees that there is another SQL pod for the same tenant in `sql_instances` system table (see https://github.com/cockroachdb/cockroach/issues/106537#issue-1797296642). Thus, it attempts to dial that SQL pod to perform the startup migration, which fails because we have `listener` not responding to that dial attempt.

**Comment by JeffSwenson** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336)_ status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)
> 
> _[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbSD8Bq7gOsLlguXVtM:bzh9mgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):_
> 
> > ```go
> > 	listener, err := net.Listen("tcp", rpcAddr)
> > 	require.NoError(t, err)
> > ```
> 
> For the failure that I got under stress on CI on this line:
> 
> ```
> === RUN   TestStartTenantWithStaleInstance
>     test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/c101b7a464a1afc1f5af0cd85792187e/logTestStartTenantWithStaleInstance3465982568
>     test_log_scope.go:81: use -show-logs to present logs inline
>     server_sql_test.go:450: 
>         	Error Trace:	github.com/cockroachdb/cockroach/pkg/ccl/serverccl/server_sql_test.go:450
>         	Error:      	Received unexpected error:
>         	            	listen tcp 127.0.0.1:45845: bind: address already in use
>         	Test:       	TestStartTenantWithStaleInstance
> ```
> 
> it seems like the stopped tenant hasn't released the socket, and to me it seems like a benign error with the test setup. I'm inclined to introduce SucceedsSoon until `net.Listen` doesn't return an error. @JeffSwenson WDYT?

Adding a `SucceedsSoon` LGTM. Another option is we could start the listener, then inject a synthetic sql_instance row. But I like the current implementation of the test because it is very generic and has minimal coupling to the values in the sql_instance table.

**Review by yuzefovich — COMMENTED** [review]
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-:-NbV7DIPDY4bsmKM3Ja1:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @JeffSwenson, @knz, and @mgartner)

___
*[`pkg/ccl/serverccl/server_sql_test.go` line 450 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/108336#-NbSD8Bq7gOsLlguXVtL:-NbV7AQxD5S5uBZ8gF5b:bcxgpet) ([raw file](https://github.com/cockroachdb/cockroach/blob/50a93086938ee580cf7b1afed1023e08974e9688/pkg/ccl/serverccl/server_sql_test.go#L450)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Adding a `SucceedsSoon` LGTM. Another option is we could start the listener, then inject a synthetic sql_instance row. But I like the current implementation of the test because it is very generic and has minimal coupling to the values in the sql_instance table.
</blockquote></details>

Thanks, let's keep `SucceedsSoon` then.

**Pull request merged**